server/src/main/scala/io/delta/standalone/internal/DeltaSharingCDCReader.scala

```diff
@@ -0,0 +1,281 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Putting these classes in this package to access Delta Standalone internal APIs
+package io.delta.standalone.internal
+
+import java.sql.Timestamp
+
+import io.delta.standalone.DeltaLog
+import io.delta.standalone.internal.actions.{
+  AddCDCFile,
+  AddFile,
+  CommitInfo,
+  FileAction,
+  Metadata,
+  RemoveFile
+}
+import io.delta.standalone.internal.exception.DeltaErrors
+import io.delta.standalone.internal.util.ConversionUtils
+import org.apache.hadoop.conf.Configuration
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ListBuffer
+
+/**
+ * A class to load Delta tables from `TableConfig`. It also caches the loaded tables internally
+ * to speed up the loading.
+ */
+class DeltaSharingCDCReader(val deltaLog: DeltaLogImpl, val conf: Configuration) {
+
+  private lazy val snapshot = deltaLog.snapshot
+  lazy val protocol = snapshot.protocolScala
+  lazy val metadata = snapshot.metadataScala
+  private lazy val history = DeltaHistoryManager(deltaLog)
+
+  private[internal] def getCDCVersions(
+      cdfOptions: Map[String, String],
+      latestVersion: Long): (Long, Long) = {
+    val startingVersion = getVersionForCDC(
+      cdfOptions,
+      DeltaDataSource.CDF_START_VERSION_KEY,
+      DeltaDataSource.CDF_START_TIMESTAMP_KEY,
+      latestVersion
+    )
+    if (startingVersion.isEmpty) {
+      throw DeltaCDFErrors.noStartVersionForCDF
+    }
+    // add a version check here that is cheap instead of after trying to list a large version
+    // that doesn't exist
+    if (startingVersion.get > latestVersion) {
+      throw DeltaCDFErrors.startVersionAfterLatestVersion(startingVersion.get, latestVersion)
+    }
+
+    val endingVersion = getVersionForCDC(
+      cdfOptions,
+      DeltaDataSource.CDF_END_VERSION_KEY,
+      DeltaDataSource.CDF_END_TIMESTAMP_KEY,
+      latestVersion
+    )
+
+    if (endingVersion.exists(_ < startingVersion.get)) {
```

Comment:
Question: if start version == end version, would we return no changes? If so, should we allow it?

Reply:
it would return changes at the version. Example, table is at version 3, and we insert 1 row and table becomes version 4, startingVersion=endingVersion=4 will return the change of inserting this 1 row.

Reply:
So [3, 4] will give the same result as [4,4] ?

Reply:
No, they are not the same. All versions are inclusive. [3,4] will return changes for version 3 as well, which might be removing 2 rows from the table.